packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts

```diff
@@ -159,6 +167,31 @@ export class AutomergeHost {
     return this._repo;
   }
 
+  private async _beforeSave({ path, batch }: BeforeSaveParams) {
+    const getDocumentIdFromPath = (path: StorageKey): DocumentId => path[0] as DocumentId;
+    const handle = this._repo.handles[getDocumentIdFromPath(path)];
+    if (!handle) {
+      return;
+    }
+    const doc = handle.docSync();
+    if (!doc) {
+      return;
+    }
+
+    const lastAvailableHash = getHeads(doc);
+
+    const getDocumentObjects = (doc: SpaceDoc): string[] => Object.keys(doc.objects ?? {});
+
+    const objectIds = getDocumentObjects(doc);
+    const encodedIds = objectIds.map((objectId) => idCodec.encode({ documentId: handle.documentId, objectId }));
+    const idToLastHash = new Map(encodedIds.map((id) => [id, lastAvailableHash]));
+    this._indexMetadataStore.markDirty(idToLastHash, batch);
+  }
+
+  private async _afterSave() {
+    this._indexMetadataStore.afterMarkDirty();
```

Comment:
In general function names should represent the actions they perform (as oposed to callbacks):

```suggestion
    this._indexMetadataStore.notifyMarkedDirty();
```

+ add jsdoc to that method